File: Soulo/Services/CognitiveDriftService.swift
import Foundation

final class CognitiveDriftService: Sendable {
    static let shared = CognitiveDriftService()

    private init() {}

    // MARK: - Drift Detection

    func detectDrift(from entries: [JournalEntry]) async throws -> DriftReport? {
        let dated = entries.compactMap { e -> (TimeInterval, VoiceBiomarkers, EmotionalState?)? in
            guard let b = e.biomarkers else { return nil }
            return (e.timestamp, b, e.emotion)
        }.sorted { $0.0 < $1.0 }

        guard dated.count >= 10 else { return nil }

        return try await Task.detached(priority: .utility) {
            let splitIndex = dated.count / 2
            let early = Array(dated.prefix(splitIndex))
            let late = Array(dated.suffix(splitIndex))

            let metrics: [(BiomarkerMetric, KeyPath<VoiceBiomarkers, Double>)] = [
                (.speechRate, \.speechRate),
                (.hesitationRate, \.hesitationRate),
                (.vocalEnergy, \.vocalEnergy),
                (.pitchInstability, \.pitchInstability),
                (.jitter, \.jitter),
                (.shimmer, \.shimmer),
            ]

            var drifts: [CognitiveDrift] = []
            var trendShiftCount = 0

            for (metric, keyPath) in metrics {
                let earlyVals = early.map { $0.1[keyPath: keyPath] }
                let lateVals = late.map { $0.1[keyPath: keyPath] }

                guard earlyVals.count >= 3, lateVals.count >= 3 else { continue }

                let earlyMean = earlyVals.reduce(0, +) / Double(earlyVals.count)
                let lateMean = lateVals.reduce(0, +) / Double(lateVals.count)

                let earlyVar = earlyVals.map { ($0 - earlyMean) * ($0 - earlyMean) }.reduce(0, +) / Double(earlyVals.count)
                let lateVar = lateVals.map { ($0 - lateMean) * ($0 - lateMean) }.reduce(0, +) / Double(lateVals.count)

                let pooledStd = sqrt((earlyVar + lateVar) / 2)
                let delta = lateMean - earlyMean

                guard pooledStd > 0 else { continue }
                let effectSize = delta / pooledStd

                let direction: TrendDirection = effectSize > 0 ? .increasing : .decreasing
                if abs(effectSize) > 0.3 { trendShiftCount += 1 }

                let significance: DriftSignificance
                switch abs(effectSize) {
                case ..<0.3: significance = .none
                case ..<0.5: significance = .mild
                case ..<0.8: significance = .moderate
                case ..<1.2: significance = .significant
                default: significance = .critical
                }

                let emotionCorr = self.correlateWithEmotion(early: early, late: late, metric: keyPath)

                drifts.append(CognitiveDrift(
                    id: UUID().uuidString,
                    windowStart: early.first!.0,
                    windowEnd: late.last!.0,
                    metric: metric,
                    delta: delta,
                    meanEarly: earlyMean,
                    meanLate: lateMean,
                    direction: direction,
                    significance: significance,
                    correlation: emotionCorr,
                    detectedAt: Date().timeIntervalSince1970
                ))
            }

            let overallSig = drifts.map { $0.significance }.max() ?? .none

            var emotionalCorr: [EmotionType: Double] = [:]
            for drift in drifts {
                for (emotion, corr) in drift.correlation {
                    emotionalCorr[emotion, default: 0] += corr
                }
            }
            if !emotionalCorr.isEmpty {
                let count = Double(drifts.filter { !$0.correlation.isEmpty }.count)
                for (key, val) in emotionalCorr { emotionalCorr[key] = val / max(count, 1) }
            }

            // Gradual vs sudden
            let timeSpan = dated.last!.0 - dated.first!.0
            let avgDaysBetweenEntries = timeSpan / Double(dated.count)
            let hasSuddenShifts = drifts.contains { abs($0.delta) > 0.5 && avgDaysBetweenEntries < 7 }
            let driftType: DriftReport.DriftType = {
                if hasSuddenShifts && trendShiftCount >= 2 { return .mixed }
                if hasSuddenShifts { return .sudden }
                return .gradual
            }()

            let summary = self.generateSummary(drifts: drifts, type: driftType)

            return DriftReport(
                id: UUID().uuidString,
                overallSignificance: overallSig,
                drifts: drifts,
                trendShiftCount: trendShiftCount,
                emotionalCorrelation: emotionalCorr,
                gradualVsSudden: driftType,
                summary: summary,
                detectedAt: Date().timeIntervalSince1970
            )
        }.value
    }

    // MARK: - Emotion Correlation

    private static func correlateWithEmotion(early: [(TimeInterval, VoiceBiomarkers, EmotionalState?)],
                                              late: [(TimeInterval, VoiceBiomarkers, EmotionalState?)],
                                              metric: KeyPath<VoiceBiomarkers, Double>) -> [EmotionType: Double] {
        let all = early + late
        let withEmotion = all.compactMap { (_, bio, emotion) -> (Double, EmotionType)? in
            guard let e = emotion else { return nil }
            return (bio[keyPath: metric], e.primaryEmotion)
        }

        var grouped: [EmotionType: [Double]] = [:]
        for (val, emotion) in withEmotion {
            grouped[emotion, default: []].append(val)
        }

        let overallMean = withEmotion.map { $0.0 }.reduce(0, +) / Double(max(withEmotion.count, 1))
        var correlations: [EmotionType: Double] = [:]
        for (emotion, vals) in grouped where vals.count >= 2 {
            let emMean = vals.reduce(0, +) / Double(vals.count)
            correlations[emotion] = (emMean - overallMean) / max(overallMean, 0.001)
        }
        return correlations
    }

    // MARK: - Summary

    private func generateSummary(drifts: [CognitiveDrift], type: DriftReport.DriftType) -> String {
        let significant = drifts.filter { $0.significance >= .moderate }
        guard !significant.isEmpty else { return "No significant cognitive drift detected." }

        let metrics = significant.map { $0.metric.rawValue.lowercased() }.joined(separator: ", ")
        let typeDesc = type == .gradual ? "gradual" : type == .sudden ? "sudden" : "mixed gradual and sudden"

        return "\(typeDesc.capitalized) cognitive drift detected in \(significant.count) metrics: \(metrics)."
    }

    // MARK: - Drift-Based Patterns

    func detectDriftPatterns(in entries: [JournalEntry]) async throws -> [DetectedPattern] {
        guard let report = try await detectDrift(from: entries) else { return [] }
        guard report.overallSignificance >= .moderate else { return [] }

        var patterns: [DetectedPattern] = []

        if report.overallSignificance >= .significant {
            let severity = report.overallSignificance.score
            patterns.append(DetectedPattern(
                patternType: .cognitiveShift,
                severity: severity,
                title: "Significant Cognitive Drift",
                message: report.summary,
                firstDetected: report.detectedAt,
                lastDetected: report.detectedAt,
                occurrenceCount: report.trendShiftCount
            ))
        }

        for drift in report.drifts where drift.significance >= .moderate {
            let severity = drift.significance.score
            let direction = drift.direction == .increasing ? "increased" : "decreased"
            patterns.append(DetectedPattern(
                patternType: .cognitiveShift,
                severity: severity,
                title: "\(drift.metric.rawValue) Shift",
                message: "Your \(drift.metric.rawValue.lowercased()) has \(direction) by \(String(format: "%.2f", abs(drift.delta))).",
                firstDetected: drift.detectedAt,
                lastDetected: drift.detectedAt,
                occurrenceCount: 1
            ))
        }

        return patterns
    }
}

File: android/app/src/main/java/com/soulo/app/services/CognitiveDriftService.kt
package com.soulo.app.services

import com.soulo.app.models.*
import kotlin.math.abs
import kotlin.math.sqrt

object CognitiveDriftService {
    private const val MIN_ENTRIES = 6
    private const val COHENS_D_THRESHOLD = 0.5

    fun detectDrift(
        entries: List<JournalEntry>,
        baseline: Map<BiomarkerMetric, Double>
    ): DriftReport? {
        val sorted = entries
            .filter { it.biomarkers != null && it.emotion != null }
            .sortedBy { it.timestamp }
        if (sorted.size < MIN_ENTRIES) return null

        val mid = sorted.size / 2
        val early = sorted.take(mid)
        val late = sorted.drop(mid)

        val drifts = mutableListOf<CognitiveDrift>()

        for ((metric, keyPath) in metricExtractors) {
            val earlyVals = early.mapNotNull { keyPath(it.biomarkers) }
            val lateVals = late.mapNotNull { keyPath(it.biomarkers) }
            if (earlyVals.size < 2 || lateVals.size < 2) continue

            val earlyMean = earlyVals.average()
            val lateMean = lateVals.average()
            val delta = lateMean - earlyMean

            val pooledStd = pooledStdDev(earlyVals, lateVals)
            val effectSize = if (pooledStd > 0) delta / pooledStd else 0.0

            val direction = when {
                effectSize > 0 -> TrendDirection.increasing
                effectSize < 0 -> TrendDirection.decreasing
                else -> TrendDirection.stable
            }

            val significance = when {
                abs(effectSize) < 0.2 -> DriftSignificance.none
                abs(effectSize) < 0.5 -> DriftSignificance.mild
                abs(effectSize) < 0.8 -> DriftSignificance.moderate
                abs(effectSize) < 1.2 -> DriftSignificance.significant
                else -> DriftSignificance.critical
            }

            val emotionCorr = correlateWithEmotion(early, late, metric)

            drifts.add(
                CognitiveDrift(
                    id = "cd_$metric",
                    windowStart = early.first().timestamp,
                    windowEnd = late.last().timestamp,
                    metric = metric,
                    delta = delta,
                    meanEarly = earlyMean,
                    meanLate = lateMean,
                    direction = direction,
                    significance = significance,
                    correlation = emotionCorr,
                    detectedAt = System.currentTimeMillis() / 1000
                )
            )
        }

        val significantDrifts = drifts.filter {
            it.significance >= DriftSignificance.moderate
        }
        if (significantDrifts.isEmpty()) return null

        val overallSig = significantDrifts.maxByOrNull { it.significance.score }?.significance
            ?: DriftSignificance.none

        val trendShiftCount = drifts.count {
            val baselineVal = baseline[it.metric] ?: it.meanEarly
            (it.meanLate - baselineVal) / (baselineVal.coerceAtLeast(0.001)) > 0.3
        }

        val emotionalCorr = mutableMapOf<EmotionType, Double>()
        var totalCorr = 0.0
        for (d in drifts) {
            for ((emotion, corr) in d.correlation) {
                emotionalCorr[emotion] = (emotionalCorr[emotion] ?: 0.0) + abs(corr)
                totalCorr += abs(corr)
            }
        }
        if (totalCorr > 0) {
            for ((emotion, value) in emotionalCorr) {
                emotionalCorr[emotion] = value / totalCorr
            }
        }

        val gradualCount = drifts.count {
            abs(it.meanLate - it.meanEarly) < COHENS_D_THRESHOLD
        }
        val driftType = when {
            gradualCount > drifts.size / 2 -> DriftReport.DriftType.gradual
            significantDrifts.size > drifts.size / 2 -> DriftReport.DriftType.sudden
            else -> DriftReport.DriftType.mixed
        }

        val summary = buildString {
            append("Detected ${significantDrifts.size} significant drift(s)")
            if (overallSig >= DriftSignificance.significant) {
                append(". Notable changes: ")
                append(significantDrifts.take(3).joinToString(", ") { it.description })
            }
        }

        return DriftReport(
            id = "drift_${sorted.last().id}",
            overallSignificance = overallSig,
            drifts = drifts,
            trendShiftCount = trendShiftCount,
            emotionalCorrelation = emotionalCorr,
            gradualVsSudden = driftType,
            summary = summary,
            detectedAt = System.currentTimeMillis() / 1000
        )
    }

    private fun pooledStdDev(a: List<Double>, b: List<Double>): Double {
        val meanA = a.average()
        val meanB = b.average()
        val varA = a.map { (it - meanA) * (it - meanA) }.average()
        val varB = b.map { (it - meanB) * (it - meanB) }.average()
        val n1 = a.size.toDouble()
        val n2 = b.size.toDouble()
        return sqrt(((n1 - 1) * varA + (n2 - 1) * varB) / (n1 + n2 - 2))
    }

    private fun correlateWithEmotion(
        early: List<JournalEntry>,
        late: List<JournalEntry>,
        metric: BiomarkerMetric
    ): Map<EmotionType, Double> {
        val correlations = mutableMapOf<EmotionType, Double>()
        val extractor = metricExtractors[metric] ?: return correlations
        val allEntries = early + late
        val lateValences = late.mapNotNull { it.emotion?.valence }

        for (emotion in EmotionType.entries) {
            val lateEmotionVals = late
                .filter { it.emotion?.primaryEmotion == emotion }
                .mapNotNull { extractor(it.biomarkers) }
            if (lateEmotionVals.size < 2) continue
            val lateBaseline = late.mapNotNull { extractor(it.biomarkers) }.average()
            if (lateBaseline == 0.0) continue
            val deviation = (lateEmotionVals.average() - lateBaseline) / lateBaseline
            correlations[emotion] = deviation.coerceIn(-1.0, 1.0)
        }

        return correlations
    }

    private val metricExtractors: Map<BiomarkerMetric, (VoiceBiomarkers?) -> Double?> = mapOf(
        BiomarkerMetric.speechRate to { it?.speechRate },
        BiomarkerMetric.vocalEnergy to { it?.vocalEnergy },
        BiomarkerMetric.pitchInstability to { it?.pitchInstability },
        BiomarkerMetric.hesitationRate to { it?.hesitationRate },
        BiomarkerMetric.microBreathCount to { it?.microBreathCount?.toDouble() },
        BiomarkerMetric.jitter to { it?.jitter },
        BiomarkerMetric.shimmer to { it?.shimmer }
    )
}
